In [1]:
import pandas as pd

TRAIN = '/kaggle/input/house-prices-advanced-regression-techniques/train.csv'

df = pd.read_csv(filepath_or_buffer=TRAIN,).drop(columns=['Id'])
df.head()

MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0          60       RL         65.0     8450   Pave   NaN      Reg   
1          20       RL         80.0     9600   Pave   NaN      Reg   
2          60       RL         68.0    11250   Pave   NaN      IR1   
3          70       RL         60.0     9550   Pave   NaN      IR1   
4          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig  ... PoolArea PoolQC Fence MiscFeature  \
0         Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
1         Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   
2         Lvl    AllPub    Inside  ...        0    NaN   NaN         NaN   
3         Lvl    AllPub    Corner  ...        0    NaN   NaN         NaN   
4         Lvl    AllPub       FR2  ...        0    NaN   NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  SalePrice  
0       0      2    2008        WD         Normal     208500  
1       0      5    2007        WD         Normal     181500  
2       0      9    2008        WD         Normal     223500  
3       0      2    2006        WD        Abnorml     140000  
4       0     12    2008        WD         Normal     250000  

[5 rows x 80 columns]

In [2]:
TEST = '/kaggle/input/house-prices-advanced-regression-techniques/test.csv'
test_df = pd.read_csv(filepath_or_buffer=TEST)

In [3]:
TARGET = 'SalePrice'
keys = [key for key, value in df.drop(columns=[TARGET]).dtypes.to_dict().items() if str(value) in {'int64', 'float64'}]
numerical_columns = [key for key in keys if test_df[key].isna().sum() == 0]

In [4]:
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

X_train, X_test, y_train, y_test = train_test_split(df[numerical_columns], df[TARGET], test_size=0.20, random_state=2024)
linreg = LinearRegression(positive=False).fit(X=X_train, y=y_train)
print('rmse: {:5.4f}'.format(mean_squared_error(squared=False, y_true=np.log(y_test), y_pred=np.log(linreg.predict(X=X_test)))))

rmse: 0.1671


In [5]:
from plotly import express
express.histogram(x=numerical_columns, y=linreg.coef_)

In [6]:
SUBMISSION = '/kaggle/working/submission.csv'

linreg_full = LinearRegression(positive=False).fit(X=df[numerical_columns], y=df[TARGET])
linreg_result_df = pd.DataFrame(data={'Id': test_df['Id'], 'SalePrice': linreg_full.predict(X=test_df[numerical_columns])})
linreg_result_df.to_csv(path_or_buf=SUBMISSION, index=False)

In [7]:
from plotly import express
express.histogram(x=numerical_columns, y=linreg_full.coef_)

In [8]:
import numpy as np
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error

ridge = Ridge(tol=1e-4, random_state=2024, max_iter=10000, positive=True, solver='lbfgs').fit(X=X_train, y=y_train)
print('rmse: {:5.4f}'.format(mean_squared_error(squared=False, y_true=np.log(y_test), y_pred=np.log(ridge.predict(X=X_test)))))

rmse: 0.1836


In [9]:
from plotly import express
express.histogram(x=ridge.feature_names_in_, y=ridge.coef_)

Our linear regressor result and our ridge result look substantially different.

In [10]:
from plotly import express

express.scatter(x=linreg_full.predict(X=test_df[numerical_columns]), y=ridge.predict(X=test_df[numerical_columns]))

In [11]:
from plotly import express
from sklearn.linear_model import Lasso

lasso = Lasso(tol=1e-4, random_state=2024, max_iter=10000, positive=False, ).fit(X=X_train, y=y_train)
print('rmse: {:5.4f}'.format(mean_squared_error(squared=False, y_true=np.log(y_test), y_pred=np.log(lasso.predict(X=X_test)))))
express.histogram(x=lasso.feature_names_in_, y=lasso.coef_).show()
express.scatter(x=linreg.predict(X=X_test), y=lasso.predict(X=X_test), log_x=True, log_y=True).show()

rmse: 0.1671


What does our delta between the linear regression result and the lasso result look like? When we plot them against each other they look essentially identical.

In [12]:
from plotly import express

express.line(y=sorted(linreg.predict(X=X_test) - lasso.predict(X=X_test))).show()

Our best model so far is Ridge with positive=True.
An RMSE of about 0.1836 on our test split corresponds to an RMSE of about 0.2201 for the test data. If we can't improve on our test RMSE we shouldn't submit a new submission.

In [13]:
from plotly import express
from sklearn.neural_network import MLPRegressor

activation = ['identity', 'logistic', 'tanh', 'relu'][3]
neural = MLPRegressor(hidden_layer_sizes=(400, 200, 100,), activation=activation, solver='adam', alpha=1e-3, batch_size='auto', 
                      learning_rate='adaptive', learning_rate_init=1e-2, power_t=0.5, max_iter=1000, shuffle=True, random_state=2024,
                      tol=1e-5, verbose=False, warm_start=False, momentum=0.8, nesterovs_momentum=True, early_stopping=False, beta_1=0.9, 
                      beta_2=0.999, epsilon=1e-08, n_iter_no_change=20, max_fun=15000).fit(X=X_train, y=y_train)
print('done after {} iterations.'.format(neural.n_iter_))
print('rmse: {:5.4f}'.format(mean_squared_error(squared=False, y_true=np.log(y_test), y_pred=np.log(neural.predict(X=X_test)))))
express.scatter(x=linreg.predict(X=X_test), y=neural.predict(X=X_test), log_x=False, log_y=False).show()

done after 259 iterations.
rmse: 0.1980


In [14]:
from plotly import express

express.line(y=sorted(linreg.predict(X=X_test) - neural.predict(X=X_test))).show()

In [15]:
SUBMISSION = '/kaggle/working/submission.csv'

neural.fit(X=df[numerical_columns], y=df[TARGET])
pd.DataFrame(data={'Id': test_df['Id'], 'SalePrice': neural.predict(X=test_df[numerical_columns])}).to_csv(path_or_buf=SUBMISSION, index=False)

A test RMSE of 0.1980 corresponds to a score of 0.2203 for this neural network regressor.

In [16]:
# ridge.fit(X=df[numerical_columns], y=df[TARGET])
# pd.DataFrame(data={'Id': test_df['Id'], 'SalePrice': ridge.predict(X=test_df[numerical_columns])}).to_csv(path_or_buf=SUBMISSION, index=False)

In [17]:
from sklearn.tree import DecisionTreeRegressor

tree = DecisionTreeRegressor(criterion='squared_error', splitter='best', max_depth=None, min_samples_split=2, min_samples_leaf=1,
                             min_weight_fraction_leaf=0.0, max_features=None, random_state=2024, max_leaf_nodes=None,
                             min_impurity_decrease=0.0, ccp_alpha=0.0, ).fit(X=X_train, y=y_train)
print('rmse: {:5.4f}'.format(mean_squared_error(squared=False, y_true=np.log(y_test), y_pred=np.log(tree.predict(X=X_test)))))
express.scatter(x=linreg.predict(X=X_test), y=tree.predict(X=X_test), log_x=False, log_y=False).show()

rmse: 0.2179


In [18]:
tree.fit(X=df[numerical_columns], y=df[TARGET])
pd.DataFrame(data={'Id': test_df['Id'], 'SalePrice': tree.predict(X=test_df[numerical_columns])}).to_csv(path_or_buf=SUBMISSION, index=False)
express.histogram(x=numerical_columns, y=tree.feature_importances_).show()